
$\qquad$ $\qquad$$\qquad$  **TDA231/DIT370 Discrete Optimization: Home Assignment 3 -- LP Duality and the Primal-Dual Algorithm** <br />
$\qquad$ $\qquad$$\qquad$                   **During grading time, direct queries regarding Q-1,2 to David & Q-3 to Marc** <br />
$\qquad$ $\qquad$$\qquad$                     **Due Date: 03/03/2025** <br />
$\qquad$ $\qquad$$\qquad$                   **Submitted by: Name, Personal No., Email** <br />



---


General guidelines:
*   All solutions to theoretical and pratical problems must be submitted in this ipynb notebook and equations, wherever required, should be formatted using LaTeX math-mode.
*   All discussion regarding practical problems, along with solutions and plots should be specified in this notebook. All plots/results should be visible such that the notebook does not have to be run. But the code in the notebook should reproduce the plots/results if we choose to do so.
*   Your name, personal number and email address should be specified above.
*   All tables and other additional information should be included in this notebook.
*   Before submitting, make sure that your code can run on another computer. That all plots can show on another computer including all your writing. It is good to check if your code can run here: https://colab.research.google.com.


# Problem 1.

Consider the following LP problem:

\begin{alignat*}{2}
\max \ &4x_1-2x_2+5x_3+6x_4+7x_5\\
\\
\textrm{s.t} \quad
&2x_1 + 2x_2 - 4x_3 + 4x_4 + 8x_5 &&\leq 6\\
&2x_1 + \ \ {}x_2 - 2x_3 - \ \ x_4 - 3x_5 &&\geq -1\\
&5x_1 - 2x_2 + 4x_3 + 4x_4 + 2x_5 &&= 5\\
&2x_1 - 2x_2 + 5x_3 + 3x_4 + \ \ x_5 &&\leq 4\\
&\hspace{5.3cm} \vec x &&\geq \vec 0
\end{alignat*}

* (4 points) Write the LP dual of this problem.
* (3 points) Use CVXPY to compute the primal and dual optimum solutions and compare their values.
* (3 points) Check the complementary slackness conditions.

In [1]:
import cvxpy as cp
import numpy as np

# Data
c = np.array([4, -2, 5, 6, 7], dtype=float)

A = np.array([
    [ 2,  2, -4,  4,  8],
    [ 2,  1, -2, -1, -3],
    [ 5, -2,  4,  4,  2],
    [ 2, -2,  5,  3,  1],
], dtype=float)

b = np.array([6, -1, 5, 4], dtype=float)

# Primal 
x = cp.Variable(5, nonneg=True)

primal_constraints = [
    A[0] @ x <= b[0],
    A[1] @ x >= b[1],
    A[2] @ x == b[2],
    A[3] @ x <= b[3],
]
primal_prob = cp.Problem(cp.Maximize(c @ x), primal_constraints)
primal_prob.solve(solver=cp.GLPK)   

print("=== PRIMAL ===")
print("status:", primal_prob.status)
print("opt value:", primal_prob.value)
print("x*:", np.array(x.value).round(6))


# Dual

y1 = cp.Variable(nonneg=True)   
y2 = cp.Variable()              
y3 = cp.Variable()              
y4 = cp.Variable(nonneg=True)   

y = cp.hstack([y1, y2, y3, y4])

dual_constraints = [
    y2 <= 0,
    A.T @ y >= c
]
dual_prob = cp.Problem(cp.Minimize(b @ y), dual_constraints)
dual_prob.solve(solver=cp.GLPK)

print("\n=== DUAL ===")
print("status:", dual_prob.status)
print("opt value:", dual_prob.value)
print("y*:", np.array(y.value).round(6))

# Compare values
print("\n=== STRONG DUALITY CHECK ===")
print("primal - dual:", float(primal_prob.value - dual_prob.value))

# Complementary slackness

xv = np.array(x.value).flatten()
yv = np.array(y.value).flatten()

# primal slacks
slack1 = b[0] - A[0] @ xv                
surplus2 = A[1] @ xv - b[1]              
slack4 = b[3] - A[3] @ xv                

# dual slacks
dual_slack = A.T @ yv - c               

print("\n=== COMPLEMENTARY SLACKNESS ===")
print("slack1 (<=):", slack1)
print("surplus2 (>=):", surplus2)
print("slack4 (<=):", slack4)
print("dual slack A^T y - c:", np.round(dual_slack, 8))

print("\nProducts y_i * slack_i (should be ~0):")
print("y1*slack1:", yv[0]*slack1)
print("y2*surplus2:", yv[1]*surplus2)
print("y4*slack4:", yv[3]*slack4)

print("\nProducts x_j * (A^T y - c)_j (should be ~0):")
print(np.round(xv * dual_slack, 10))


=== PRIMAL ===
status: optimal
opt value: 9.220338983050846
x*: [0.627119 2.813559 1.542373 0.       0.661017]

=== DUAL ===
status: optimal
opt value: 9.220338983050848
y*: [ 0.525424 -0.372881  0.338983  1.      ]

=== STRONG DUALITY CHECK ===
primal - dual: -1.7763568394002505e-15

=== COMPLEMENTARY SLACKNESS ===
slack1 (<=): 8.881784197001252e-16
surplus2 (>=): 6.661338147750939e-16
slack4 (<=): 8.881784197001252e-16
dual slack A^T y - c: [ 0.          0.         -0.          0.83050847  0.        ]

Products y_i * slack_i (should be ~0):
y1*slack1: 4.666700171305743e-16
y2*surplus2: -2.4838888008562817e-16
y4*slack4: 8.881784197001252e-16

Products x_j * (A^T y - c)_j (should be ~0):
[ 0.  0. -0.  0.  0.]


Problem 1: Primal/Dual results, strong duality, and complementary slackness

Primal optimum
The solver returns an optimal primal solution  
$x^\star = (0.627119,\; 2.813559,\; 1.542373,\; 0,\; 0.661017)$  
with objective value  
$c^T x^\star = 9.220338983050846$.  
Status: optimal.


Dual optimum
The solver returns an optimal dual solution  
$y^\star = (y_1,y_2,y_3,y_4) = (0.525424,\; -0.372881,\; 0.338983,\; 1)$  
with objective value  
$b^T y^\star = 9.220338983050848$.  
Status: optimal.

Strong duality
The objective values match up to numerical tolerance:  
$c^T x^\star - b^T y^\star \approx -1.7763568394\cdot 10^{-15} \approx 0$.  

Therefore strong duality holds (the primal and dual optimal values are equal).


Complementary slackness
Primal constraint slacks (from the solver output):**
- Constraint 1 (a $\le$ constraint): $\text{slack}_1 \approx 8.88\cdot 10^{-16}\approx 0$
- Constraint 2 (a $\ge$ constraint), using surplus $(a_2^T x^\star - b_2)$: $\text{surplus}_2 \approx 6.66\cdot 10^{-16}\approx 0$
- Constraint 4 (a $\le$ constraint): $\text{slack}_4 \approx 8.88\cdot 10^{-16}\approx 0$
(Constraint 3 is an equality and is tight by definition.)

The products are approximately zero:
- $y_1\cdot \text{slack}_1 \approx 0$
- $y_2\cdot \text{surplus}_2 \approx 0$
- $y_4\cdot \text{slack}_4 \approx 0$

Dual slack:
The solver reports  
$A^T y^\star - c = (0,\;0,\;0,\;0.83050847,\;0)$.

Complementary slackness for variables requires  
$x_j^\star\big((A^T y^\star)_j - c_j\big)=0$ for each $j$,  
and the computed products are all (numerically) zero.

In particular, $(A^T y^\star)_4 - c_4 = 0.83050847 > 0$ implies $x_4^\star = 0$, which matches the primal solution.



Conclusion
Both primal and dual solve to optimality, the optimal values agree (strong duality), and complementary slackness holds up to numerical tolerance. This confirms that the reported solutions $x^\star$ and $y^\star$ are consistent and correct.


# Problem 2.

Consdier the LP problem:
\begin{alignat*}{2}
\max \ &6x_1 - 5x_3\\
\\
\textrm{s.t} \quad
&6x_1 - 3x_2 + x_3 &&= 2\\
&3x_1 + 4x_2 + x_3 &&\leq 5\\
&\ \ x_1 - 7x_2 &&\leq 5\\
&\hspace{2.45cm} x_1 &&\geq 0\\
&\hspace{2.45cm} x_2 &&\leq 0\\
&\hspace{2.45cm} x_3 &&\text{ unrestricted}
\end{alignat*}

* (3 points) Write the LP dual of this problem.
* (4 points) Consider the feasible solution $\vec x = (0,0,2)$ to the primal. Check if this solution is optimal by using the complementary slackness conditions to write down the corresponding dual solution.
* (3 points) Use complementary slackness to check if the primal feasible solution $\vec x = (1,0,-4)$ is optimal.

## Problem 2 Duality and Complementary Slackness 

### Primal (given)
Maximize $6x_1-5x_3$ subject to  
(1) $6x_1-3x_2+x_3=2$  
(2) $3x_1+4x_2+x_3\le 5$  
(3) $x_1-7x_2\le 5$  
and variable signs $x_1\ge 0,\ x_2\le 0,\ x_3$ free.


### (a) Dual LP
Let $y_1$ correspond to (1), $y_2$ to (2), $y_3$ to (3).  
Because (1) is an equality, $y_1$ is free; because (2) and (3) are $\le$ constraints in a maximization problem, $y_2\ge 0$ and $y_3\ge 0$.

Dual objective:
$ \min\ 2y_1+5y_2+5y_3.$

Dual constraints come from columns and the sign rules for primal variables:
- since $x_1\ge 0$: $(A^Ty)_1\ge 6$,
- since $x_2\le 0$: $(A^Ty)_2\le 0$,
- since $x_3$ is free: $(A^Ty)_3 = -5$.

The dual is:
- Minimize $2y_1+5y_2+5y_3$ subject to  
$6y_1+3y_2+y_3\ge 6$  
$-3y_1+4y_2-7y_3\le 0$  
$y_1+y_2=-5$  
$y_2\ge 0,\ y_3\ge 0,\ y_1$ free.

### (b) Check $\bar x=(0,0,2)$ with complementary slackness
First, $\bar x$ is primal-feasible: it satisfies (1) and both inequalities.

Compute slacks of the $\le$ constraints:
$s_2 = 5-(3x_1+4x_2+x_3)=5-2=3>0$,  
$s_3 = 5-(x_1-7x_2)=5-0=5>0$.

Complementary slackness for $\le$ constraints gives $y_2s_2=0$ and $y_3s_3=0$, hence
$y_2=0$ and $y_3=0$.  
Then the dual equality $y_1+y_2=-5$ forces $y_1=-5$.

So CS forces the dual candidate $(y_1,y_2,y_3)=(-5,0,0)$, but it is not dual feasible because
$6y_1+3y_2+y_3=6(-5)=-30 \not\ge 6$.

Therefore, $\bar x$ is feasible but not optimal.


### (c) Check $\tilde x=(1,0,-4)$ with complementary slackness
$\tilde x$ is primal-feasible.

Slacks:
$s_2 = 5-(3\cdot 1+4\cdot 0+(-4)) = 5-(-1)=6>0$,  
$s_3 = 5-(1-7\cdot 0)=5-1=4>0$.

Again CS implies $y_2=0$ and $y_3=0$, hence $y_1=-5$.  
But since $x_1=1>0$ and the dual constraint for $x_1$ is $6y_1+3y_2+y_3\ge 6$, complementary slackness on variables requires this constraint to be tight:
$6y_1+3y_2+y_3 = 6$.
With $(-5,0,0)$ the left-hand side is $-30\neq 6$, a contradiction.

Therefore, $\tilde x$ is feasible but not optimal.



### Find the true optimum and an optimal dual
From (1): $x_3=2-6x_1+3x_2$. Substitute into the objective:
$6x_1-5x_3 = 6x_1-5(2-6x_1+3x_2)=36x_1-15x_2-10$.

Let $y=-x_2\ge 0$ (since $x_2\le 0$). Constraint (3) becomes $x_1+7y\le 5$, and the objective becomes
$36x_1+15y-10$.
Over the triangle $x_1\ge 0,\ y\ge 0,\ x_1+7y\le 5$, the maximum occurs at $(x_1,y)=(5,0)$.
Thus $x_2=0$ and $x_3=2-6\cdot 5=-28$.

So the primal optimum is
$x^\star=(5,0,-28)$ with value $6\cdot 5-5(-28)=170$.

A matching dual feasible solution is $y^\star=(-5,0,36)$:
- $y_1+y_2=-5$ holds,
- $6y_1+3y_2+y_3 = -30+36=6$,
- $-3y_1+4y_2-7y_3 = 15-252=-237\le 0$,
- $y_2\ge 0,\ y_3\ge 0$.
Dual value: $2(-5)+5\cdot 0+5\cdot 36=170$, matching the primal value.


# Problem 3.

Consider the primal-dual algorithm for vertex cover discussed in class.
* (4 points) Run the algorithm by hand on the graph in the figure below (from your previous homework). Show the values of the primal and dual variables at each iteration.
* (6 points) Implement the primal-dual algorithm as a python script to compute (approximate) vertex covers and run it for the random graph $G(100,0.1)$ from the previous homework.

<img src="https://tinyurl.com/tsnuz2c" alt="Drawing" style="width: 180px;"/>

If the image does not load, try the direct link: https://tinyurl.com/tsnuz2c

## Problem 3  Primal–Dual Algorithm for Vertex Cover

### LP relaxation and dual
For a graph $G=(V,E)$ with vertex weights $w(v)>0$, the LP relaxation of the minimum weight vertex cover problem is

Primal (P):  
$\min \sum_{v\in V} w(v)x_v$  

subject to  
$x_u + x_v \ge 1$ for every $(u,v)\in E$, and $x_v \ge 0$ for all $v\in V$.

The corresponding dual is

Dual (D):  
$\max \sum_{e\in E} y_e$  

subject to  
$\sum_{e\ni v} y_e \le w(v)$ for all $v\in V$, and $y_e \ge 0$ for all $e\in E$.

The primal dual algorithm starts with $y_e=0$ for all edges, residuals $r(v)=w(v)$ for all vertices, and an empty cover $S=\emptyset$.  
While there exists an uncovered edge $(u,v)$ with $u,v\notin S$, I increase its dual variable by  
$\Delta=\min\{r(u),r(v)\}$,  
update the residuals, and add any vertex that becomes tight $r(v)=0$ to $S$.  
A useful interpretation is that the algorithm builds a feasible dual solution and a feasible vertex cover at the same time. The standard analysis gives  
$w(S)\le 2\sum_{e\in E}y_e \le 2\cdot \mathrm{}$.


## (a) Run the algorithm by hand on the given graph

The numbers inside the nodes are the weights. We label the vertices by position:
- $a$ top-left, $w(a)=4$
- $b$ top-right, $w(b)=2$
- $c$ middle-left, $w(c)=3$
- $d$ middle-right, $w(d)=2$
- $e$ bottom-left, $w(e)=1$
- $f$ bottom-right, $w(f)=4$

Edges (as drawn): $(a,b)$, $(a,c)$, $(c,e)$, $(b,d)$, $(c,d)$, $(d,f)$, $(a,d)$.

Initialization:
All dual variables $y_e=0$.  
Residuals: $r(a)=4,r(b)=2,r(c)=3,r(d)=2,r(e)=1,r(f)=4$.  
Cover $S=\emptyset$ (so $x_v=0$ for all vertices).

### Iteration 1: choose edge $(c,e)$
$\Delta=\min\{r(c),r(e)\}=\min\{3,1\}=1$.  
Update: $y_{ce}=1$, $r(c)=2$, $r(e)=0$.  
Vertex $e$ becomes tight $\Rightarrow S=\{e\}$ (so $x_e=1$).

### Iteration 2: choose edge $(b,d)$
$\Delta=\min\{r(b),r(d)\}=\min\{2,2\}=2$.  
Update: $y_{bd}=2$, $r(b)=0$, $r(d)=0$.  
Vertices $b$ and $d$ become tight $\Rightarrow S=\{e,b,d\}$.

### Iteration 3: remaining uncovered edge $(a,c)$
$\Delta=\min\{r(a),r(c)\}=\min\{4,2\}=2$.  
Update: $y_{ac}=2$, $r(a)=2$, $r(c)=0$.  
Vertex $c$ becomes tight $\Rightarrow S=\{e,b,d,c\}$.

Now every edge has at least one endpoint in $S$, so we stop.

**Final primal (cover):** $S=\{b,c,d,e\}$ with cost  
$w(S)=w(b)+w(c)+w(d)+w(e)=2+3+2+1=8$.

**Final dual variables (nonzero):** $y_{ce}=1$, $y_{bd}=2$, $y_{ac}=2$, so  
$\sum_{e\in E} y_e = 1+2+2=5$.

Check the bound: $w(S)=8 \le 2\cdot 5=10$.



## (b) Implementation and run on $G(100,0.1)$

We generated a random graph $G(100,0.1)$ with a fixed seed (seed $=0$) and ran the primal dual algorithm for the unweighted case ($w(v)=1$ for all $v$). The run produced:

- $|E| = 520$
- Cover size $|S| = 79$
- Dual value $\sum_e y_e = 47$
- $2\sum_e y_e = 94$
- The output set is a valid vertex cover.
- The inequality $|S|\le 2\sum_e y_e$ holds ($79\le 94$), consistent with the 2-approximation guarantee.


In [3]:
import numpy as np

# Generate random graph G(n,p)
def gen_gnp(n=100, p=0.1, seed=0):
    rng = np.random.default_rng(seed)
    E = []
    for i in range(n):
        r = rng.random(n - i - 1)
        js = np.where(r < p)[0] + i + 1
        for j in js:
            E.append((i, int(j)))
    return E

def is_vertex_cover(S, E):
    S = np.asarray(S, dtype=int)
    for (u, v) in E:
        if S[u] == 0 and S[v] == 0:
            return False
    return True
# Primal–dual algorithm 
def primal_dual_vc(E, n, w=None, eps=1e-12, reverse_delete=True):
    if w is None:
        w = np.ones(n, dtype=float)
    w = np.array(w, dtype=float)

    y = np.zeros(len(E), dtype=float)   
    r = w.copy()                        
    S = np.zeros(n, dtype=int)          

    def find_uncovered():
        for k, (u, v) in enumerate(E):
            if S[u] == 0 and S[v] == 0:
                return k, u, v
        return None

    while True:
        pick = find_uncovered()
        if pick is None:
            break
        k, u, v = pick
        delta = min(r[u], r[v])

        if delta <= eps:

            S[u] = 1
            continue

        y[k] += delta
        r[u] -= delta
        r[v] -= delta

        if r[u] <= eps:
            r[u] = 0.0
            S[u] = 1
        if r[v] <= eps:
            r[v] = 0.0
            S[v] = 1


    if reverse_delete:
        verts = np.where(S == 1)[0].tolist()
        for v in verts:
            S[v] = 0
            if not is_vertex_cover(S, E):
                S[v] = 1

    return S, y

# Run on G(100,0.1)
n = 100
p = 0.1
seed = 0

E = gen_gnp(n=n, p=p, seed=seed)
w = np.ones(n)  

S, y = primal_dual_vc(E, n, w=w, reverse_delete=True)

cover_size = int(S.sum())
dual_value = float(y.sum())

print(f"G(n,p) with n={n}, p={p}, seed={seed}")
print(f"|E| = {len(E)}")
print("Cover size |S| =", cover_size)
print("Dual value sum(y_e) =", dual_value)
print("2 * dual value =", 2 * dual_value)
print("Is vertex cover? ->", is_vertex_cover(S, E))
print("Check |S| <= 2*sum(y):", cover_size <= 2 * dual_value + 1e-6)


G(n,p) with n=100, p=0.1, seed=0
|E| = 520
Cover size |S| = 79
Dual value sum(y_e) = 47.0
2 * dual value = 94.0
Is vertex cover? -> True
Check |S| <= 2*sum(y): True
